srml/contract/src/gas.rs

```diff
@@ -35,11 +38,50 @@ impl GasMeterResult {
 	}
 }
 
+#[cfg(not(test))]
+pub trait TestAuxiliaries {}
+#[cfg(not(test))]
+impl<T> TestAuxiliaries for T {}
+
+#[cfg(test)]
+pub trait TestAuxiliaries: Any + Debug + PartialEq + Eq {}
+#[cfg(test)]
+impl<T: Any + Debug + PartialEq + Eq> TestAuxiliaries for T {}
+
+/// This trait represents a token that can be used for charging `GasMeter`.
+/// There is no other way of charging it.
+///
+/// Implementing type is expected to be super lightweight hence `Copy` (`Clone` is added
+/// for consistency). If inlined there should be no observable difference compared
+/// to a hand-written code.
+pub trait Token<T: Trait>: Copy + Clone + TestAuxiliaries {
+	/// Metadata type, which the token can require for calculating the amount
+	/// of gas to charge. Can be a some configuration type or
+	/// just the `()`.
+	type Metadata;
+
+	/// Calculate amount of gas that should be taken by this token.
+	///
+	/// Returns `None` if the amount can't be calculated e.g. because of overflow.
+	/// This situation is treated as if out of gas happened.
+	fn calculate_amount(&self, metadata: &Self::Metadata) -> Option<T::Gas>;
```

Comment:
Reading the docs, I feel this should be `Result<T::Gas, Self::Error>` rather than an Option. However, seeing the way this actually be used and is implemented, one could also argue that the docs should instead say "... by this token or None if calculation failed, which consumes all gas".

I am still in favor of having an explicit error case with a specific Type "OutOfGas" that the GasMeter can handle directly but actually properly fails for the others to make this a more flexible and cleaner model... 

I leave this decision to you.

Reply:
This function can't actually fail. I'd expect that implementors of this function will always do some basic branching and arithmetics, and there should be no heavy operations like accessing to the storage. The only problem that I can foresee here is an overflow.

I wanted to communicate it by returning an `Option` not `Result` deliberately.

But your post gave me a thought. Maybe we don't want to introduce a special mode of failing and instead of `None` we can document that people can use `T::Gas::max_value` or saturating operations to signal the overflow? The gas meter will consume all gas anyway (aka "out of gas happened") if it receives the request for gas more than available.

Reply:
That sounds more expressive and I'd prefer that!

Reply:
I would not use `max_value` to express some special value. We have `Option`, so use it. Makes the code much clearer, than starting to match on magical numbers.

Reply:
@bkchr the idea is not to use `max_value` to match against it (as in `if x == max_value() { .. }`), but to use the fact that there can't be more `max_value` thus consuming `max_value` units of gas means consuming all available gas... 

So on the implementor side I expect for code like this:

```rust
let code_len_in_gas = <T::Gas as As<u64>>::sa(self.0);
let cost = metadata
    .put_code_per_byte_cost
    .checked_mul(&code_len_in_gas)?;

Some(cost)
```

to be like this

```rust
let code_len_in_gas = <T::Gas as As<u64>>::sa(self.0);
let cost = metadata
    .put_code_per_byte_cost
    .saturating_mul(&code_len_in_gas); // if this overflows then `max_value` is returned

cost
```

Also worth noting the fact that it is already possible to do that right now, i.e. you can return result of `saturating_mul` wrapped in `Some` and get the same effect as if you returned `None`.

Reply:
Okay, ty :)